In [2]:
#Adding required packages
using Pkg
Pkg.add("CSV")
Pkg.add("DataFrames")

    Updating registry at `C:\Users\anoop\.julia\registries\General.toml`
   Resolving package versions...
   Installed SentinelArrays ── v1.4.5
   Installed FilePathsBase ─── v0.9.22
   Installed WeakRefStrings ── v1.4.2
   Installed Tables ────────── v1.12.0
   Installed InlineStrings ─── v1.4.2
   Installed PooledArrays ──── v1.4.3
   Installed WorkerUtilities ─ v1.6.1
   Installed CSV ───────────── v0.10.14
    Updating `C:\Users\anoop\.julia\environments\v1.10\Project.toml`
  [336ed68f] + CSV v0.10.14
    Updating `C:\Users\anoop\.julia\environments\v1.10\Manifest.toml`
  [336ed68f] + CSV v0.10.14
  [e2d170a0] + DataValueInterfaces v1.0.0
  [48062228] + FilePathsBase v0.9.22
  [842dd82b] + InlineStrings v1.4.2
  [82899510] + IteratorInterfaceExtensions v1.0.0
  [2dfb63ee] + PooledArrays v1.4.3
  [91c51154] + SentinelArrays v1.4.5
  [3783bdb8] + TableTraits v1.0.1
  [bd369af6] + Tables v1.12.0
  [ea10d353] + WeakRefStrings v1.4.2
  [76eceee3] + WorkerUtilities v1.6.1
  [9fa8497b] + 

In [3]:
#Loading the dataset:StudentPerformanceFactors.csv
using CSV, DataFrames
df = CSV.File("StudentPerformanceFactors.csv") |> DataFrame

6607×20 DataFrame
  Row │ Hours_Studied  Attendance  Parental_Involvement  Access_to_Resources   ⋯
      │ Int64          Int64       String7               String7               ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │            23          84  Low                   High                  ⋯
    2 │            19          64  Low                   Medium
    3 │            24          98  Medium                Medium
    4 │            29          89  Low                   Medium
    5 │            19          92  Medium                Medium                ⋯
    6 │            19          88  Medium                Medium
    7 │            29          84  Medium                Low
    8 │            25          78  Low                   High
    9 │            17          94  Medium                High                  ⋯
   10 │            23          98  Medium                Medium
   11 │            17          80  Low                   High
  ⋮   │       ⋮            ⋮                ⋮                     ⋮            ⋱
 6598 │            16          91  High                  Medium
 6599 │             9          87  Low                   Medium                ⋯
 6600 │            30          94  Medium                Low
 6601 │            12          98  Medium                Low
 6602 │            20          83  Medium                Low
 6603 │            25          69  High                  Medium                ⋯
 6604 │            23          76  High                  Medium
 6605 │            20          90  Medium                Low
 6606 │            10          86  High                  High
 6607 │            15          67  Medium                Low                   ⋯
                                                16 columns and 6586 rows omitted

In [4]:
#Summary statistics for numerical columns
describe(df)

20×7 DataFrame
 Row │ variable                    mean     min       median  max           nm ⋯
     │ Symbol                      Union…   Any       Union…  Any           In ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ Hours_Studied               19.9753  1         20.0    44               ⋯
   2 │ Attendance                  79.9774  60        80.0    100
   3 │ Parental_Involvement                 High              Medium
   4 │ Access_to_Resources                  High              Medium
   5 │ Extracurricular_Activities           No                Yes              ⋯
   6 │ Sleep_Hours                 7.02906  4         7.0     10
   7 │ Previous_Scores             75.0705  50        75.0    100
   8 │ Motivation_Level                     High              Medium
   9 │ Internet_Access                      No                Yes              ⋯
  10 │ Tutoring_Sessions           1.49372  0         1.0     8
  11 │ Family_Income                        High              Medium
  12 │ Teacher_Quality                      High              Medium
  13 │ School_Type                          Private           Public           ⋯
  14 │ Peer_Influence                       Negative          Positive
  15 │ Physical_Activity           2.96761  0         3.0     6
  16 │ Learning_Disabilities                No                Yes
  17 │ Parental_Education_Level             College           Postgraduate     ⋯
  18 │ Distance_from_Home                   Far               Near
  19 │ Gender                               Female            Male
  20 │ Exam_Score                  67.2357  55        67.0    101
                                                               2 columns omitted

In [17]:
#Finding the unique values in some columns
unique_values = unique(df[!,:Parental_Education_Level])

#Finding unique values over all the columns
unique_values_dict = Dict()

for col in names(df)
    unique_values = unique(df[!, col])
    unique_values_dict[col] = unique_values
end

for (col,values) in unique_values_dict
    println("Unique values for $col:",values)
end

Unique values for Access_to_Resources:String7["High", "Medium", "Low"]
Unique values for Physical_Activity:[3, 4, 2, 1, 5, 0, 6]
Unique values for Motivation_Level:String7["Low", "Medium", "High"]
Unique values for Peer_Influence:String15["Positive", "Negative", "Neutral"]
Unique values for Internet_Access:String3["Yes", "No"]
Unique values for Attendance:[84, 64, 98, 89, 92, 88, 78, 94, 80, 97, 83, 82, 68, 60, 70, 75, 99, 74, 65, 62, 91, 90, 66, 69, 72, 63, 61, 86, 77, 71, 67, 87, 73, 96, 100, 81, 95, 79, 85, 76, 93]
Unique values for Distance_from_Home:String15["Near", "Moderate", "Far"]
Unique values for Parental_Education_Level:String15["High School", "College", "Postgraduate"]
Unique values for Sleep_Hours:[7, 8, 6, 10, 9, 5, 4]
Unique values for School_Type:String7["Public", "Private"]
Unique values for Teacher_Quality:String7["Medium", "High", "Low"]
Unique values for Learning_Disabilities:String3["No", "Yes"]
Unique values for Hours_Studied:[23, 19, 24, 29, 25, 17, 21, 9, 10, 1

In [12]:
#To find the missing values, the following code is performed:
filter(row -> any(ismissing, row), df)

229×20 DataFrame
 Row │ Hours_Studied  Attendance  Parental_Involvement  Access_to_Resources  E ⋯
     │ Int64          Int64       String7               String7              S ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │            14          60  High                  Medium               N ⋯
   2 │            17          97  Medium                Medium               N
   3 │            15          87  Low                   Medium               N
   4 │            23          82  Low                   Medium               Y
   5 │            24          90  Low                   Low                  N ⋯
   6 │            15          64  Medium                Medium               Y
   7 │            15          70  High                  Medium               N
   8 │            10          80  High                  Medium               Y
   9 │            20          66  High                  High                 N ⋯
  10 │            21          65  High                  High                 N
  11 │            21          91  High                  Medium               N
  ⋮  │       ⋮            ⋮                ⋮                     ⋮             ⋱
 220 │            29          75  Low                   Medium               Y
 221 │            11          72  High                  Low                  N ⋯
 222 │            12          89  Medium                High                 Y
 223 │            14          77  Medium                Medium               Y
 224 │            26          77  Low                   Medium               Y
 225 │            23          64  Medium                Medium               N ⋯
 226 │             9          84  Medium                Medium               N
 227 │            22          90  Low                   High                 N
 228 │             9          90  High                  High                 Y
 229 │            17          92  Medium                Medium               N ⋯
                                                 16 columns and 208 rows omitted

There are 229 rows which has some missing values. Since the dataset is so large, let's omit all these rows with missing values for better handling the data.

In [15]:
# Remove rows with missing values
df = dropmissing(df)

6378×20 DataFrame
  Row │ Hours_Studied  Attendance  Parental_Involvement  Access_to_Resources   ⋯
      │ Int64          Int64       String7               String7               ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │            23          84  Low                   High                  ⋯
    2 │            19          64  Low                   Medium
    3 │            24          98  Medium                Medium
    4 │            29          89  Low                   Medium
    5 │            19          92  Medium                Medium                ⋯
    6 │            19          88  Medium                Medium
    7 │            29          84  Medium                Low
    8 │            25          78  Low                   High
    9 │            17          94  Medium                High                  ⋯
   10 │            23          98  Medium                Medium
   11 │            17          80  Low                   High
  ⋮   │       ⋮            ⋮                ⋮                     ⋮            ⋱
 6369 │            16          91  High                  Medium
 6370 │             9          87  Low                   Medium                ⋯
 6371 │            30          94  Medium                Low
 6372 │            12          98  Medium                Low
 6373 │            20          83  Medium                Low
 6374 │            25          69  High                  Medium                ⋯
 6375 │            23          76  High                  Medium
 6376 │            20          90  Medium                Low
 6377 │            10          86  High                  High
 6378 │            15          67  Medium                Low                   ⋯
                                                16 columns and 6357 rows omitted

Now the dataframe df is clean without any missing values. Now, let's begin analyzing the data.

I will be going through some statistical summaries and visualizations first, to understand the data better.

I'm using StatsBase package, which is a powerful tool for statistical analysis.

In [23]:
Pkg.add("StatsBase")
using StatsBase

# Get statistics for a specific column (e.g., 'Hours_Studied')
# To round the figure, we can use thr round function.

mean_hours = round(mean(df.Hours_Studied),digits=3)
std_hours = round(std(df.Hours_Studied), digits=3)

println("Mean Hours Studied: ", mean_hours)
println("Standard Deviation of Hours Studied: ", std_hours)


   Resolving package versions...
  No Changes to `C:\Users\anoop\.julia\environments\v1.10\Project.toml`
  No Changes to `C:\Users\anoop\.julia\environments\v1.10\Manifest.toml`


Mean Hours Studied: 19.977
Standard Deviation of Hours Studied: 5.985
